feat(settings): categorize user settings

tthvo opened the pull request:
# Welcome to Cryostat! 👋
## Before contributing, make sure you have:
* [x] Read the [contributing guidelines](https://github.com/cryostatio/cryostat/blob/main/CONTRIBUTING.md)
* [x] Linked a relevant issue which this PR resolves
* [x] Linked any other relevant issues, PR's, or documentation, if any
* [x] Resolved all conflicts, if any
* [x] Rebased your branch PR on top of the latest upstream `main` branch
* [x] Attached at least one of the following labels to the PR: `[chore, ci, docs, feat, fix, test]`
* [x] Signed the last commit: `git commit --amend --signoff`
_______________________________________________

Fixes: #781 
Depends on #789 

## Description of the change:

- [x] Categorize user settings (similar to OCP console). Settings can be ordered based using an order index.
- [x] Enable flip and add height constraints for setting menus.
- [x] Fix nested forms, anchors warnings.
- [ ] Tests for setting page.

## Motivation for the change:

See #781 

## How to manually test:
1. Run `CRYOSTAT_IMAGE=quay.io... sh smoketest.sh..`.
2. Go to setting page to see new changes.

## Screenshots
![new-setting-page](https://user-images.githubusercontent.com/68053619/211077579-d896f2be-4c56-423c-87e6-000514a6f2f5.png)

tthvo commented:
What do you think? I will start on writing tests if this could be accepted :D Just unfortunately, allowing flip on menu causes the snapshots to fail so we need to remove them.

maxcao13 commented:
I like it, it's very nice! I don't know if this is just me, but I seem to like the `Dashboard` coming first before `Notification and Messages` b/c of alphabetical order but because `General` and `Advanced` are "special" cases, I would leave those.

Also, is it just me or did the `DEVELOPMENT` dropdown in the Feature Level disappear?

tthvo commented:
> I don't know if this is just me, but I seem to like the Dashboard coming first before Notification and Messages b/c of alphabetical order but because General and Advanced are "special" cases, I would leave those.

Thanks I was thinking of putting setting groups that might be more frequently visited to the top. `Language & Region` is also defined but hidden. I will show it when #789 is merged.

tthvo commented:
> Also, is it just me or did the DEVELOPMENT dropdown in the Feature Level disappear?

Still seeing it tho...

![Screenshot from 2023-01-06 14-13-38](https://user-images.githubusercontent.com/68053619/211083174-da6aedbd-ab01-4f96-9fe3-c6490d8c34a5.png)

andrewazores commented:
![image](https://user-images.githubusercontent.com/3787464/211095966-abba869b-9e82-488c-83cb-b35553e7f94c.png)

Development setting seems to be missing for me too.

andrewazores commented:
I tried in different browsers and in private/incognito modes as well to avoid any effects where it could be from some change to the localstorage format, but in all cases it was the same.

andrewazores commented:
Anyway other than that I think this looks great.

tthvo commented:
Ahh I remember it now. `DEVELOPMENT` is hidden is production build. This is intended right? 

https://github.com/cryostatio/cryostat-web/blob/181e70efe076710c35ab56ac187212095db5ef29/src/app/Settings/FeatureLevels.tsx#L83

andrewazores commented:
Ha. Right, good catch. Didn't even think of that when I pulled the GH pkgs PR preview image.

Looks great. Still need to add more tests?

tthvo commented:
> Looks great. Still need to add more tests?

Oh yess! I gonna take this chance to add tests for setting panels. Also need to rebase and  move language picker to setting. I will put it as draft for now.

tthvo commented:
Ready for review again....Changes since last comments:

- [x] Move language select to setting page + comment out `zh` field so that its not shown as an option.
- [x] Wrap setting components in `FeatureFlag` to hide and show base on current feature level. 
  - [x] Except for `PRODUCTION`-leveled settings, others will have a compact label next to its title (try Language setting).
  - [x] Only show tabs with at least 1 setting that satisfy current feature level.
 - [x] Expand setting page to fill available vertical space (avoid jumps and make it similar to OCP console). 
 - [x]  Add unit tests for setting components.

Got a bit issue with circular deps in `Settings.test.tsx` tests, so I disable eslint import for that file for now.


Screenshots:

![latest-design](https://user-images.githubusercontent.com/68053619/211558767-6a4957b6-a6f7-402d-861b-564cbbb48bd2.png)

maxcao13 commented on src/app/Settings/AutomatedAnalysisConfig.tsx:
```diff
@@ -108,4 +108,5 @@ export const AutomatedAnalysisConfig: UserSetting = {
   description:
     'Set the recording configuration for automated analysis recordings. You may want smaller or larger values for max-age and max-size depending on how recent you want events to be recorded from the analysis.',
   content: Component,
+  category: 'Dashboard',
```
Maybe instead of hardcording category names, there should be enum somewhere of possible categories for better typing and safeguarding?

e.g. 
`DashboardCategories.DASHBOARD = 'Dashboard'`

Also can some of these hard-coded strings start to be put into the localization files? At the least the new ones that we are introducing.

The workflow is:
1. const [t] = useTranslation();
2. (in this case): 
```js
category: t('Dashboard', {ns: 'common'}); //since this seems to be a string we will use commonly throughout Cryostat.
```
3. Then use `yarn localize` to generate the translation key in the `locales/en/common.js` files
4. Edit the json file key to whatever you wanted to hardcode.

tthvo replied:
> Maybe instead of hardcording category names, there should be enum somewhere of possible categories for better typing and safeguarding?

Ahh for this I was using string union type so only those strings can be used. Just a thought that we are having a lot of circular dependencies so I am just avoiding importing enums.

tthvo replied:
> Also can some of these hard-coded strings start to be put into the localization files? At the least the new ones that we are introducing.

Opps. Totally forgot to add after rebasing XD Adding now.

tthvo replied:
Changes are becoming quite large to track so I will open a follow-up PR after some more readings and tryouts with alternatives to dynamic keys.

maxcao13 reviewed (commented):
Otherwise, looks good!

andrewazores commented on src/app/AppLayout/AppLayout.tsx:
```diff
@@ -337,28 +332,6 @@ const AppLayout: React.FC<AppLayoutProps> = ({ children }) => {
     ),
     [handleHelpToggle]
   );
-  const handleLanguageToggle = React.useCallback(() => setShowLanguageDropdown((v) => !v), [setShowLanguageDropdown]);
```
I think it adds some visual clutter but it is much more discoverable there for sure.

Maybe adding an item to the user dropdown menu that just links to the relevant Settings view makes sense? It seems like a pretty common pattern to put the language selection under the user menu so it should be fairly familiar and discoverable.

andrewazores replied:
This does mean that the Basic auth form is going to be stuck in English for the user's first login, though.

The `/settings` route currently requires the user to be logged in to visit which should not be the case, so perhaps if we fix that and also add some link from the Basic auth form component to the Settings language selection then that will work. That is also the one place that Patternfly does have a design guideline for including a language selection.

tthvo replied:
Ahh I was removing it as it seems to make the top-right action toolbar clustered. What do you think @andrewazores?

tthvo replied:
Ohh opps I was a bit late on commenting XD

tthvo replied:
I found Patternfly suggest a language selection on login form here: https://www.patternfly.org/v4/components/login-page/design-guidelines/#elements. 

But i guess we still do want to make `/settings` accessible without login?

tthvo replied:
![Screenshot from 2023-01-10 13-29-01](https://user-images.githubusercontent.com/68053619/211632896-dd784f6d-a387-4b97-ad0f-145907321961.png)

tthvo replied:
![Screenshot from 2023-01-10 13-34-19](https://user-images.githubusercontent.com/68053619/211633897-f71af605-7a23-4e52-813e-0fdff104d570.png)

tthvo replied:
> But i guess we still do want to make /settings accessible without login?

Ah nvmind just saw new PR. Connection is pretty slow today :D nothing on github today is updating on time haha XD

andrewazores commented on src/app/Settings/FeatureLevels.tsx:
```diff
@@ -99,4 +99,6 @@ export const FeatureLevels: UserSetting = {
   description: 'Control which graphical features appear in the application.',
   content: Component,
+  category: 'General',
+  orderInGroup: 1,
```
I think this should go under `Advanced` actually.

andrewazores replied:
If this is moved to `Advanced` then the remaining two items under `General` are both seldom-used options relating to WebSockets, so perhaps `General` should be renamed to `WebSockets` or maybe `Connectivity`.

tthvo replied:
I removed `General` for now until we have some general settings later :D

tthvo commented:
Just also notice that if users are currently on `/settings` and log out there, that setting page is persisted since it is anonymous. I added a patch to redirect back to `/` so that login form is shown. What do you think?

maxcao13 reviewed (approved):
Looks good, and code is nice!

tthvo commented on src/app/Shared/Services/Login.service.tsx:
```diff
@@ -254,7 +254,8 @@ export class LoginService {
     this.removeCacheItem(this.AUTH_METHOD_KEY);
-    window.location.href = window.location.href.split('#')[0];
+    const url = new URL(window.location.href.split('#')[0]);
+    window.location.href = url.pathname.match(/\/settings/i) ? '/' : url.pathname;
```
Ahh yes I did check with OpenShift with these changes and the flow remains unaffected (i.e. i think we are reloading at the end of logout anyway).

Anw sorry launching a working cluster took is quite a pain lately :D took forever XD
